src/cs3iface.py

```diff
@@ -9,137 +9,215 @@
+  if filedownloadres.status.code == cs3code.CODE_NOT_FOUND:
+    ctx['log'].info('msg="File not found on read" filepath="%s"' % filepath)
+    yield IOError('No such file or directory')
+  elif filedownloadres.status.code != cs3code.CODE_OK:
+    ctx['log'].debug('msg="Failed to initiateFileDownload on read" filepath="%s" reason="%s"' % \
+                     (filepath, filedownloadres.status.message))
+    yield IOError(filedownloadres.status.message)
+  ctx['log'].debug('msg="readfile: InitiateFileDownloadRes returned" endpoint="%s"' % filedownloadres.download_endpoint)
+
+  # Download
   try:
-    tstart = time.clock()
-    f = open(_getfilename(filename), mode='wb')
-    tend = time.clock()
-    log.info('msg="File open for write" filename="%s" elapsedTimems="%.1f"' % (filename, (tend-tstart)*1000))
-    # write the file. In a future implementation, we should find a way to only update the required chunks...
-    written = f.write(content)
-    f.close()
-    if written != size:
-      raise IOError('Written %d bytes but content is %d bytes' % (written, size))
-  except OSError as e:
-    log.warning('msg="Error writing to file" filename="%s" error="%s"' % (filename, e))
-    raise IOError(e)
-
-def renamefile(_endpoint, origfilename, newfilename, ruid, rgid):
-  '''Rename a file from origfilename to newfilename on behalf of the given uid, gid.'''
-  try:
-    os.rename(_getfilename(origfilename), _getfilename(newfilename))
-  except (FileNotFoundError, PermissionError) as e:
-    raise IOError(e)
-
-def removefile(_endpoint, filename, _ruid, _rgid, _force=0):
-  '''Remove a file on behalf of the given uid, gid.
-     The force argument is irrelevant and ignored for local storage.'''
+    fileget = requests.get(url=filedownloadres.download_endpoint,
+                           headers={'x-access-token': _authenticate(userid)})
+  except requests.exceptions.RequestException as e:
+    ctx['log'].error('msg="Exception when downloading file from Reva" reason="%s"' % e)
+    yield IOError(e)
+  tend = time.clock()
+  data = fileget.content
+  if fileget.status_code != http.client.OK:
+    ctx['log'].error('msg="Error downloading file from Reva" code="%d" reason="%s"' % (fileget.status_code, fileget.reason))
+    yield IOError(fileget.reason)
+  else:
+    ctx['log'].info('msg="File open for read" filepath="%s" elapsedTimems="%.1f"' % (filepath, (tend-tstart)*1000))
+    for i in range(0, len(data), ctx['chunksize']):
+      yield data[i:i+ctx['chunksize']]
+
+
+def writefile(_endpoint, filepath, userid, content, _noversion=0):
+  '''Write a file using the given userid as access token. The entire content is written
+    and any pre-existing file is deleted (or moved to the previous version if supported).
+    The noversion flag is currently not supported.'''
+  tstart = time.clock()
+  # prepare endpoint
+  req = sp.InitiateFileUploadRequest(ref=spr.Reference(path=filepath))
+  fileuploadres = ctx['cs3stub'].InitiateFileUpload(
```

Comment:
This is not a fileuploadres, let's rename this.

Reply:
It's a response of a FileUpload request, similarly to the readflie case. I'm fine with any renaming you wish.

Reply:
Not really since it's not the actual file upload, I'll rename.